shell/platform/android/test/io/flutter/plugin/editing/InputConnectionAdaptorTest.java

```diff
@@ -0,0 +1,250 @@
+
+        assertTrue(didConsume);
+        assertEquals(selStart, Selection.getSelectionStart(editable));
+        assertEquals(selStart - 1, Selection.getSelectionEnd(editable));
+    }
+
+    @Test
+    public void testSendKeyEvent_rightKeyMovesCaretRight() {
+        int selStart = 5;
+        Editable editable = sampleEditable(selStart, selStart);
+        InputConnectionAdaptor adaptor = sampleInputConnectionAdaptor(editable);
+
+        KeyEvent rightKeyDown = new KeyEvent(KeyEvent.ACTION_DOWN, KeyEvent.KEYCODE_DPAD_RIGHT);
+        boolean didConsume = adaptor.sendKeyEvent(rightKeyDown);
+
+        assertTrue(didConsume);
+        assertEquals(selStart + 1, Selection.getSelectionStart(editable));
+        assertEquals(selStart + 1, Selection.getSelectionEnd(editable));
+    }
+
+    @Test
+    public void testSendKeyEvent_rightKeyExtendsSelectionRight() {
+        int selStart = 5;
+        int selEnd = 40;
+        Editable editable = sampleEditable(selStart, selEnd);
+        InputConnectionAdaptor adaptor = sampleInputConnectionAdaptor(editable);
+
+        KeyEvent rightKeyDown = new KeyEvent(KeyEvent.ACTION_DOWN, KeyEvent.KEYCODE_DPAD_RIGHT);
+        boolean didConsume = adaptor.sendKeyEvent(rightKeyDown);
+
+        assertTrue(didConsume);
+        assertEquals(selStart, Selection.getSelectionStart(editable));
+        assertEquals(selEnd + 1, Selection.getSelectionEnd(editable));
+    }
+
+    @Test
+    public void testSendKeyEvent_shiftRightKeyStartsSelectionRight() {
+        int selStart = 5;
+        Editable editable = sampleEditable(selStart, selStart);
+        InputConnectionAdaptor adaptor = sampleInputConnectionAdaptor(editable);
+
+        KeyEvent shiftRightKeyDown = new KeyEvent(0, 0, KeyEvent.ACTION_DOWN, KeyEvent.KEYCODE_DPAD_RIGHT, 0, KeyEvent.META_SHIFT_ON);
+        boolean didConsume = adaptor.sendKeyEvent(shiftRightKeyDown);
+
+        assertTrue(didConsume);
+        assertEquals(selStart, Selection.getSelectionStart(editable));
+        assertEquals(selStart + 1, Selection.getSelectionEnd(editable));
+    }
+
+    @Test
+    public void testSendKeyEvent_upKeyMovesCaretUp() {
+        int selStart = SAMPLE_TEXT.indexOf('\n') + 4;
+        Editable editable = sampleEditable(selStart, selStart);
+        InputConnectionAdaptor adaptor = sampleInputConnectionAdaptor(editable);
+
+        KeyEvent upKeyDown = new KeyEvent(KeyEvent.ACTION_DOWN, KeyEvent.KEYCODE_DPAD_UP);
+        boolean didConsume = adaptor.sendKeyEvent(upKeyDown);
+
+        assertTrue(didConsume);
+        int expectedSelection = selStart - SAMPLE_TEXT.indexOf('\n');
+        // WIP
+        // The asserts below don't work. KEYCODE_DPAD_UP moves the caret to
+        // index 0, as if the text were in one single line.
+        // assertEquals(expectedSelection, Selection.getSelectionStart(editable));
+        // assertEquals(expectedSelection, Selection.getSelectionEnd(editable));
```

Comment:
Couldn't get up/down to work as expected yet.

Up moves selection to beginning of the text and down moves to end. There's a line break in the text so it should move to the next/previous line instead.

Not sure why it behaves differently in tests.. Will look more tomorrow.

Reply:
So I have noticed that depending on the input method, some keystrokes/events do not pass through this method at all. I am not yet clear on where those events go but that behavior may have something to do with this.